wdl-format/src/format.rs

```diff
@@ -0,0 +1,912 @@
+    }
+    result.push_str(struct_def.name().as_str());
+    result.push_str(&format_inline_comment(
+        &SyntaxElement::Token(struct_def.name().syntax().clone()),
+        false,
+    ));
+
+    let open_brace = struct_def
+        .syntax()
+        .children_with_tokens()
+        .find(|c| c.kind() == SyntaxKind::OpenBrace)
+        .expect("struct definition should have an open brace");
+    result.push_str(&format_preceding_comments(
+        &open_brace,
+        0,
+        !result.ends_with(NEWLINE),
+    ));
+    if !result.ends_with(NEWLINE) {
+        result.push(' ');
+    }
+    result.push('{');
+    result.push_str(&format_inline_comment(&open_brace, true));
+
+    for decl in struct_def.members() {
+        result.push_str(&format_declaration(&Decl::Unbound(decl), 1));
+    }
+
+    let close_brace = struct_def
+        .syntax()
+        .children_with_tokens()
+        .find(|c| c.kind() == SyntaxKind::CloseBrace)
+        .expect("struct definition should have a close brace");
+    result.push_str(&format_preceding_comments(&close_brace, 0, false));
+    result.push('}');
+    result.push_str(&format_inline_comment(&close_brace, true));
+
+    result
+}
+
+/// Format a WDL document.
+pub fn format_document(code: &str) -> Result<String, Vec<Diagnostic>> {
+    let (document, diagnostics) = Document::parse(code);
+    if !diagnostics.is_empty() {
+        return Err(diagnostics);
+    }
+    let mut validator = Validator::default();
+    match validator.validate(&document) {
+        Ok(_) => {
+            // The document is valid, so we can format it.
+        }
+        Err(diagnostics) => return Err(diagnostics),
+    }
+
+    let mut result = String::new();
+    result.push_str(&format_version_statement(
+        document.version_statement().unwrap(),
+    ));
+    result.push_str(NEWLINE);
+
+    let ast = document.ast();
+    let ast = ast.as_v1().unwrap();
+    result.push_str(&format_imports(ast.imports()));
+
+    ast.items().for_each(|item| {
+        match item {
+            DocumentItem::Import(_) => {
+                // Imports have already been formatted
+            }
+            DocumentItem::Workflow(workflow_def) => {
+                if !result.ends_with(&NEWLINE.repeat(2)) {
+                    result.push_str(NEWLINE);
+                }
+                result.push_str(&format_workflow(&workflow_def));
+            }
+            DocumentItem::Task(task_def) => {
+                if !result.ends_with(&NEWLINE.repeat(2)) {
+                    result.push_str(NEWLINE);
+                }
+                result.push_str(&format_task(&task_def));
+            }
+            DocumentItem::Struct(struct_def) => {
+                if !result.ends_with(&NEWLINE.repeat(2)) {
+                    result.push_str(NEWLINE);
+                }
+                result.push_str(&format_struct_definition(&struct_def));
+            }
+        };
+    });
+
+    Ok(result)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests also need something more scalable—akin to what we do in the other packages where files are more easily observed and reasoned about (rather than embedding the text right in the rust source).

Reply:
Totally agree. That was on my agenda to deal with tomorrow.